File: src/main/kotlin/club/kazza/kazzacraft/utils/NibbleArray.java
package club.kazza.kazzacraft.utils;

public final class NibbleArray {
    private final byte[] backing;

    public NibbleArray(int size) {
        backing = new byte[size / 2];
    }

    public byte[] getBacking() {
        return backing;
    }

    public byte get(int index) {
        byte val = backing[index / 2];
        if (index % 2 == 0) {
            return (byte) (val & 0x0f);
        } else {
            return (byte) ((val & 0xf0) >> 4);
        }
    }

    public void set(int index, int value) {
        value &= 0xf;
        int half = index / 2;
        byte previous = backing[half];
        if (index % 2 == 0) {
            backing[half] = (byte) (previous & 0xf0 | value);
        } else {
            backing[half] = (byte) (previous & 0x0f | value << 4);
        }
    }
}

File: crafty-common/src/main/kotlin/world/crafty/common/utils/NibbleArray.kt
package world.crafty.common.utils

class NibbleArray(val backing: ByteArray) {
    constructor(size: Int) : this(ByteArray(size/2))

    operator fun get(index: Int) : Int {
        val value = backing[index / 2].toInt()
        return if(index % 2 == 0) (value and 0x0f) else ((value and 0xf0) shr 4)
    }

    operator fun set(index: Int, value: Int) {
        val nibble = value and 0xf
        val halfIndex = index / 2
        val previous = backing[halfIndex].toInt()
        if(index % 2 == 0) {
            backing[halfIndex] = ((previous and 0xf0) or nibble).toByte()
        } else {
            backing[halfIndex] = ((previous and 0x0f) or (nibble shl 4)).toByte()
        }
    }
    
    val size = backing.size * 2
}